File: app/src/main/java/com/zacharee1/systemuituner/services/SafeModeService.java
package com.zacharee1.systemuituner.services;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.database.ContentObserver;
import android.net.Uri;
import android.os.Build;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.preference.PreferenceManager;
import android.provider.Settings;

import com.zacharee1.systemuituner.R;
import com.zacharee1.systemuituner.activites.settings.SettingsActivity;
import com.zacharee1.systemuituner.util.SettingsUtils;

public class SafeModeService extends Service {
    private ShutDownReceiver mShutDownReceiver;
    private ThemeChangeReceiver mThemeReceiver;

    private ResolutionChangeListener mResListener;

    private Handler mHandler;
    private ContentObserver observer;
    private SharedPreferences preferences;

    public SafeModeService() {
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        mHandler = new Handler(Looper.getMainLooper());
        preferences = PreferenceManager.getDefaultSharedPreferences(this);

        startInForeground();
        restoreStateOnStartup();
        restoreQSHeaderCount();
        restoreQSRowColCount();
        setUpReceivers();
        setUpContentObserver();
        restoreSnoozeState();
        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        try {
            unregisterReceiver(mThemeReceiver);
        } catch (Exception e) {}

        try {
            unregisterReceiver(mShutDownReceiver);
        } catch (Exception e) {}

        try {
            getContentResolver().unregisterContentObserver(mResListener);
        } catch (Exception e) {}

        try {
            getContentResolver().unregisterContentObserver(observer);
        } catch (Exception e) {}
    }

    private void startInForeground() {
        PendingIntent settingsIntent = PendingIntent.getActivity(this, 0, new Intent(this, SettingsActivity.class), 0);
        NotificationManager manager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);

        Notification.Builder notification = new Notification.Builder(this)
                .setSmallIcon(R.mipmap.ic_launcher)
                .setContentTitle(getResources().getString(R.string.notif_title))
                .setContentText(getResources().getString(R.string.notif_desc))
                .setPriority(Notification.PRIORITY_MIN)
                .setContentIntent(settingsIntent);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel("zacharee1", "SystemUI Tuner", NotificationManager.IMPORTANCE_LOW);
            manager.createNotificationChannel(channel);
            notification.setChannelId("zacharee1");
        }

        startForeground(1001, notification.build());
    }

    private void setUpReceivers() {
        mShutDownReceiver = new ShutDownReceiver();
        mThemeReceiver = new ThemeChangeReceiver();
        mResListener = new ResolutionChangeListener(mHandler);
    }

    private void restoreStateOnStartup() {
        String blacklist = Settings.Secure.getString(getContentResolver(), "icon_blacklist");

        if (blacklist == null || blacklist.isEmpty()) {
            String blacklistBackup = Settings.Global.getString(getContentResolver(), "icon_blacklist_backup");

            if (blacklistBackup != null && !blacklistBackup.isEmpty()) {
                SettingsUtils.writeSecure(this, "icon_blacklist", blacklistBackup);
            }
        }

        String qsAnimState = Settings.Secure.getString(getContentResolver(), "sysui_qs_fancy_anim");

        if (qsAnimState == null || qsAnimState.isEmpty() || qsAnimState.equals("1")) {
            String backupState = Settings.Global.getString(getContentResolver(), "sysui_qs_fancy_anim_backup");

            if (backupState != null && !backupState.isEmpty()) {
                SettingsUtils.writeSecure(this, "sysui_qs_fancy_anim", backupState);
            }
        }

        SettingsUtils.writeGlobal(this, "system_booted", "1");
    }

    private void restoreQSHeaderCount() {
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.M) {
            int count = preferences.getInt("qs_header_count", -1);
            if (count != -1) SettingsUtils.writeSecure(this, "sysui_qqs_count", String.valueOf(count));
        }
    }

    private void saveQSHeaderCount() {
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.M) {
            int count = Settings.Secure.getInt(getContentResolver(), "sysui_qqs_count", -1);
            if (count != -1) preferences.edit().putInt("qs_header_count", count).apply();
        }
    }

    private void saveQSRowColCount() {
        int row = Settings.Secure.getInt(getContentResolver(), "qs_tile_row", -1);
        int col = Settings.Secure.getInt(getContentResolver(), "qs_tile_column", -1);
        if (row != -1) preferences.edit().putInt("qs_tile_row", row).apply();
        if (col != -1) preferences.edit().putInt("qs_tile_column", col).apply();
    }

    private void restoreQSRowColCount() {
        int row = preferences.getInt("qs_tile_row", -1);
        int col = preferences.getInt("qs_tile_column", -1);

        if (row != -1) SettingsUtils.writeSecure(this, "qs_tile_row", row + "");
        if (col != -1) SettingsUtils.writeSecure(this, "qs_tile_column", col + "");
    }

    private void restoreSnoozeState() {
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.O) {
            String saved = preferences.getString("notification_snooze_options", "");

            if (!saved.isEmpty()) {
                SettingsUtils.writeGlobal(this, "notification_snooze_options", saved);
            }
        }
    }

    private void saveSnoozeState() {
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.O) {
            String set = Settings.Global.getString(getContentResolver(), "notification_snooze_options");
            if (set != null && !set.isEmpty()) preferences.edit().putString("notification_snooze_options", set).apply();
        }
    }

    private void setUpContentObserver() {
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.M) {
            observer = new ContentObserver(new Handler(Looper.getMainLooper())) {
                @Override
                public void onChange(boolean selfChange, Uri uri) {
                    if (uri.equals(Settings.Secure.getUriFor("sysui_qqs_count"))) {
                        restoreQSHeaderCount();
                    } else if (uri.equals(Settings.Secure.getUriFor("qs_tile_row")) || uri.equals(Settings.Secure.getUriFor("qs_tile_column"))) {
                        restoreQSRowColCount();
                    }
                }
            };

            getContentResolver().registerContentObserver(Settings.Global.CONTENT_URI, true, observer);
        }
    }

    private void resetBlacklist(boolean restore) {
        final String blacklist = Settings.Secure.getString(getContentResolver(), "icon_blacklist");

        SettingsUtils.writeGlobal(this, "icon_blacklist_backup", blacklist);
        SettingsUtils.writeSecure(this, "icon_blacklist", "");

       if (restore) {
           mHandler.postDelayed(new Runnable() {
               @Override
               public void run() {
                   SettingsUtils.writeSecure(SafeModeService.this, "icon_blacklist", blacklist);
               }
           }, 400);
       }
    }

    public class ShutDownReceiver extends BroadcastReceiver {
        public ShutDownReceiver() {
            IntentFilter filter = new IntentFilter();
            filter.addAction(Intent.ACTION_SHUTDOWN);
            filter.addAction(Intent.ACTION_REBOOT);
            filter.addAction("android.intent.action.QUICKBOOT_POWEROFF");
            filter.addAction("com.htc.intent.action.QUICKBOOT_POWEROFF");

            registerReceiver(this, filter);
        }

        @Override
        public void onReceive(Context context, Intent intent) {
            resetBlacklist(false);
            saveQSHeaderCount();
            saveQSRowColCount();
            saveSnoozeState();
        }
    }

    public class ThemeChangeReceiver extends BroadcastReceiver {
        public ThemeChangeReceiver() {
            IntentFilter filter = new IntentFilter("broadcast com.samsung.android.theme.themecenter.THEME_APPLY");

            registerReceiver(this, filter);
        }

        @Override
        public void onReceive(Context context, Intent intent) {
            resetBlacklist(true);
        }
    }

    public class ResolutionChangeListener extends ContentObserver {
        public ResolutionChangeListener(Handler handler) {
            super(handler);

            getContentResolver().registerContentObserver(Settings.Secure.CONTENT_URI, true, this);
        }

        @Override
        public void onChange(boolean selfChange, Uri uri) {
            Uri twRes = Settings.Secure.getUriFor("default_display_size_forced");
            Uri res = Settings.Secure.getUriFor("display_size_forced");

            if (uri.equals(twRes) || uri.equals(res)) {
                resetBlacklist(true);
            }
        }
    }
}

File: app/src/main/java/com/zacharee1/systemuituner/services/SafeModeService.kt
package com.zacharee1.systemuituner.services

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.*
import android.database.ContentObserver
import android.net.Uri
import android.os.Build
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import android.provider.Settings
import android.view.Surface
import android.view.WindowManager
import androidx.core.app.NotificationCompat
import com.zacharee1.systemuituner.R
import com.zacharee1.systemuituner.util.*
import com.zacharee1.systemuituner.util.PrefManager.Companion.AUDIO_SAFE
import com.zacharee1.systemuituner.util.PrefManager.Companion.HIGH_BRIGHTNESS_WARNING
import com.zacharee1.systemuituner.util.PrefManager.Companion.NOTIFICATION_SNOOZE_OPTIONS
import com.zacharee1.systemuituner.util.PrefManager.Companion.QQS_COUNT
import com.zacharee1.systemuituner.util.PrefManager.Companion.QS_FANCY_ANIM
import com.zacharee1.systemuituner.util.PrefManager.Companion.SAFE_MODE_NOTIF
import com.zacharee1.systemuituner.util.PrefManager.Companion.TILE_COLUMN
import com.zacharee1.systemuituner.util.PrefManager.Companion.TILE_ROW

class SafeModeService : Service() {
    private val shutDownReceiver by lazy { ShutDownReceiver() }
    private val themeReceiver by lazy { ThemeChangeReceiver() }
    private val resListener by lazy { ResolutionChangeListener(handler) }

    private val handler: Handler = Handler(Looper.getMainLooper())
    private val observer: ContentObserver = object : ContentObserver(handler) {
        override fun onChange(selfChange: Boolean, uri: Uri) {
            if (uri == Settings.Secure.getUriFor(QQS_COUNT)) {
                restoreQSHeaderCount()
            } else if (uri == Settings.Secure.getUriFor(TILE_ROW)
                    || uri == Settings.Secure.getUriFor(TILE_COLUMN)) {
                restoreQSRowColCount()
            } else if (uri == Settings.Global.getUriFor(NOTIFICATION_SNOOZE_OPTIONS)) {
                restoreSnoozeState()
            }
        }
    }

    private val prefsListener: SharedPreferences.OnSharedPreferenceChangeListener = SharedPreferences.OnSharedPreferenceChangeListener { preferences, key ->
        when (key) {
            SAFE_MODE_NOTIF -> {
                val on = preferences.getBoolean(key, true)
                if (!on) {
                    stopForeground(true)
                } else {
                    startInForeground()
                }
            }
        }
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun onCreate() {
        super.onCreate()

        startInForeground()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        prefs.registerOnSharedPreferenceChangeListener(prefsListener)

        restoreStateOnStartup()
        restoreQSHeaderCount()
        restoreHBWState()
        restoreQSRowColCount()
        setUpContentObserver()
        restoreSnoozeState()
        restoreVolumeWarning()
        return Service.START_STICKY
    }

    override fun onDestroy() {
        super.onDestroy()

        try {
            unregisterReceiver(themeReceiver)
        } catch (e: Exception) {
        }

        try {
            unregisterReceiver(shutDownReceiver)
        } catch (e: Exception) {
        }

        try {
            contentResolver.unregisterContentObserver(resListener)
        } catch (e: Exception) {
        }

        try {
            contentResolver.unregisterContentObserver(observer)
        } catch (e: Exception) {
        }

        try {
            prefs.unregisterOnSharedPreferenceChangeListener(prefsListener)
        } catch (e: Exception) {
        }
    }

    private fun startInForeground() {
        if (prefs.safeModeNotif) {
            val settingsIntent = PendingIntent.getActivity(this, 0, getNotificationSettingsForChannel("safe_mode_2"), 0)
            val manager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                val channel = NotificationChannel("safe_mode_2", resources.getString(R.string.safe_mode), NotificationManager.IMPORTANCE_LOW)
                manager.createNotificationChannel(channel)
            }

            val notification = NotificationCompat.Builder(this, "safe_mode_2")
                    .setSmallIcon(R.mipmap.ic_launcher)
                    .setContentTitle(resources.getString(R.string.notif_title))
                    .setContentText(resources.getString(R.string.notif_desc))
                    .setPriority(NotificationCompat.PRIORITY_MIN)
                    .setContentIntent(settingsIntent)

            startForeground(1001, notification.build())
        }
    }

    private fun restoreStateOnStartup() {
        restoreBlacklist()
        restoreFancyAnim()
    }

    private fun restoreHBWState() {
        if (prefs.safeModeHbw && checkSamsung()) {
            writeSystem(HIGH_BRIGHTNESS_WARNING, if (prefs.highBrightnessWarning) 0 else 1, false)
        }
    }

    private fun restoreBlacklist() {
        if (prefs.safeModeStatusBar) {
            val blacklist = blacklistManager.currentBlacklist

            if (blacklist.isEmpty()) {
                val blacklistBackup = blacklistManager.backupBlacklist

                if (blacklistBackup.isNotEmpty()) {
                    blacklistManager.setCurrentBlacklist(blacklistBackup)
                }
            }
        }
    }

    private fun restoreFancyAnim() {
        if (prefs.safeModeFancyAnim) {
            writeSecure(QS_FANCY_ANIM, if (prefs.qsFancyAnim) 1 else 0)
        }
    }

    private fun restoreVolumeWarning() {
        if (prefs.safeModeVolumeWarning) {
            writeGlobal(AUDIO_SAFE, if (prefs.audioSafe) 3 else 2)
        }
    }

    private fun restoreQSHeaderCount() {
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.M && prefs.safeModeHeaderCount) {
            val count = prefs.qqsCount
            if (count != -1) writeSecure(QQS_COUNT, count)
        }
    }

    private fun restoreQSRowColCount() {
        val rotation = (getSystemService(Context.WINDOW_SERVICE) as WindowManager).defaultDisplay.rotation
        if (prefs.safeModeRowCol) {
            if (rotation != Surface.ROTATION_180 && rotation != Surface.ROTATION_270) {
                val row = prefs.tileRow
                val col = prefs.tileColumn

                if (row != -1) writeSecure(PrefManager.TILE_ROW, row)
                if (col != -1) writeSecure(PrefManager.TILE_COLUMN, col)
            } else {
                val row = prefs.tileRowLandscape
                val col = prefs.tileColumnLandscape

                if (row != -1) writeSecure(PrefManager.TILE_ROW, row)
                if (col != -1) writeSecure(PrefManager.TILE_COLUMN, col)
            }
        }
    }

    private fun restoreSnoozeState() {
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.O && prefs.safeModeSnoozeOptions) {
            val saved = prefs.notificationSnoozeOptions

            if (saved?.isEmpty() == false) {
                writeGlobal(NOTIFICATION_SNOOZE_OPTIONS, saved)
            }
        }
    }

    private fun saveSnoozeState() {
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.O && prefs.safeModeSnoozeOptions) {
            val set = Settings.Global.getString(contentResolver, NOTIFICATION_SNOOZE_OPTIONS)
            if (set != null && !set.isEmpty()) prefs.notificationSnoozeOptions = set
        }
    }

    private fun setUpContentObserver() {
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.M) {
            contentResolver.registerContentObserver(Settings.Global.CONTENT_URI, true, observer)
            contentResolver.registerContentObserver(Settings.Secure.CONTENT_URI, true, observer)
        }
    }

    private fun resetBlacklist(restore: Boolean) {
        if (prefs.safeModeStatusBar) {
            blacklistManager.apply {
                backupBlacklist = currentBlacklist
                setCurrentBlacklist(null)

                if (restore) {
                    handler.postDelayed({ currentBlacklist = backupBlacklist }, 400)
                }
            }
        }
    }

    inner class ShutDownReceiver : BroadcastReceiver() {
        init {
            val filter = IntentFilter()
            filter.addAction(Intent.ACTION_SHUTDOWN)
            filter.addAction(Intent.ACTION_REBOOT)
            filter.addAction("android.intent.action.QUICKBOOT_POWEROFF")
            filter.addAction("com.htc.intent.action.QUICKBOOT_POWEROFF")

            registerReceiver(this, filter)
        }

        override fun onReceive(context: Context, intent: Intent) {
            resetBlacklist(false)
            saveSnoozeState()
        }
    }

    inner class ThemeChangeReceiver : BroadcastReceiver() {
        init {
            val filter = IntentFilter("com.samsung.android.theme.themecenter.THEME_APPLY")

            registerReceiver(this, filter)
        }

        override fun onReceive(context: Context, intent: Intent) {
            resetBlacklist(true)
        }
    }

    inner class ResolutionChangeListener(handler: Handler?) : ContentObserver(handler) {
        init {

            contentResolver.registerContentObserver(Settings.Secure.CONTENT_URI, true, this)
        }

        override fun onChange(selfChange: Boolean, uri: Uri) {
            val twRes = Settings.Secure.getUriFor("default_display_size_forced")
            val res = Settings.Secure.getUriFor("display_size_forced")

            if (uri == twRes || uri == res) {
                resetBlacklist(true)
            }
        }
    }
}
